# DAT-510 Network Security and  reliability
## Asahi  Cantu Moreno
## 253964

# Part II. Simplified DES

In this section, you will implement a simplified version of the DES block cipher algorithm. Naturally
enough, it is called SDES, and it is designed to have the features of the DES algorithm but scaled down
so it is more tractable to understand. (Note however, that SDES is in no way secure and should not
be used for serious cryptographic applications.)

* SDES encryption takes a 10 bit raw key (from which two 8 bit keys are generated as described in the
handout)
* Encrypts an 8 bit plaintext to produce an 8 bit ciphertext.

#### Important:
**SDES Algorithm is implemented as class SDES in file SDES.py**

In [1]:
from SDES import SDES
import time

### Tests
* To verify that your implementation of SDES is correct, try the following test cases:

| Raw Key    | PlainText  | CipherText  |
| ---        | --- | --- |
| 0000000000 | 10101010  | 00010001  |
| 1110001110 | 10101010  | 11001010  |
| 1110001110 | 01010101  | 01110000  |
| 1111111111 | 10101010  | 00000100  |
---


In [2]:
keys = ['0000000000',
        '1110001110',
        '1110001110',
        '1111111111']

texts = ['10101010',
        '10101010',
        '01010101',
        '10101010']
        

ciphers = [
    '00010001',
    '11001010',
    '01110000',
    '00000100']


sdes = SDES()
for i in range(len(keys)):
    cipher = sdes.encrypt(texts[i],keys[i])
    print(f'{keys[i]} {texts[i]}  = {cipher} == {ciphers[i]}?  {cipher == ciphers[i]}')

0000000000 10101010  = 00010001 == 00010001?  True
1110001110 10101010  = 11001010 == 11001010?  True
1110001110 01010101  = 01110000 == 01110000?  True
1111111111 10101010  = 00000100 == 00000100?  True


In [3]:
keys = [
    '0000000000',
    '0000011111',
    '0010011111',
    '0010011111',
    '1111111111',
    '1111111111',
    '1000101110',
    '1000101110']

texts = [
    '10101010',
    '11111100',
    '10100101',
    '01010101',
    None,
    None,
    None,
    None]

ciphers = [
    None,
    None,
    None,
    None,
    '00001111',
    '01000011',
    '00011100',
    '11000010']

for i in range(len(keys)):
    if texts[i] is None:
        texts[i] = sdes.decrypt(ciphers[i],keys[i])
    else:
        ciphers[i] = sdes.encrypt(texts[i],keys[i])
print('Raw key\tPlainText\tCipherText')
for i in range(len(keys)):
    print(f'{keys[i]}\t{texts[i]}\t{ciphers[i]}')


Raw key	PlainText	CipherText
0000000000	10101010	00010001
0000011111	11111100	10011101
0010011111	10100101	10010000
0010011111	01010101	11001100
1111111111	11111111	00001111
1111111111	01100001	01000011
1000101110	00111000	00011100
1000101110	00001100	11000010


## Task1.
Use your implementation to complete the following table:

| Raw Key    | PlainText  | CipherText |
| ---        |:---:       |:---:       |
| 0000000000 | 10101010  | 00010001  |
| 0000011111 | 11111100  | 10011101  |
| 0010011111 | 10100101  | 10010000  |
| 0010011111 | 01010101  | 11001100  |
| 1111111111 | 11111111  | 00001111  |
| 1111111111 | 01100001  | 01000011  |
| 1000101110 | 00111000  | 00011100  |
| 1000101110 | 00001100  | 11000010  |

---

The DES algorithm uses keys of length 56 bits, which, when DES was originally designed, was thought
to be secure enough to meet most needs. However, due to Moores law, the increase in computing
power makes it more tractible to brute-force crack a 56-bit key. Thus, an alternative version of DES
using longer keys was desirable. The result, known as Triple DES uses two 56-bit raw keys k1 and k2
and is implemented by composing DES with itself three times in the following way

$Enc_{3DES}(p)=Enc_{DES}(k_1,Dec_{DES}(k_2,Enc_{DES}(k_1,p)))$

Here, p is the paintext to encrypt, $Enc_{DES}$ is the usuarl DES encryption algorithm and $Dec_{DES}$ is the DES encryption algorithm. This strategy doubles the number of bits in the key, at the expense of performing three times as many calculations.

The tripleDES decryption algorithm is just the reverse:

$Dec_{3DES(c)}=Dec_{DES}(k_1,Enc_{DES}(k_2,Dec_{DES}(k_1,c)))$


In [4]:
keys = [
['1000101110','0110101110'],
['1000101110','0110101110'],
['1111111111','1111111111'],
['0000000000','0000000000'],
['1000010110','0110101110'],
['1011101111','0110101110'],
['1111111111','1111111111'],
['0000000000','0000000000']]

texts = [
    '11010111',
    '10101010',
    '00000000',
    '01010010',
    None,
    None,
    None,
    None]


ciphers = [
    None,
    None,
    None,
    None,
    '11100110',
    '01010000',
    '00000100',
    '11110000']

for i in range(len(keys)):
    if texts[i] is None:
        texts[i] = sdes.decrypt3SDES(ciphers[i],keys[i][0],keys[i][1])
    else:
        ciphers[i] = sdes.encrypt3SDES(texts[i],keys[i][0],keys[i][1])
print('Raw key 1\tRaw key 2\tPlainText\tCipherText')
for i in range(len(keys)):
    print(f'{keys[i][0]}\t{keys[i][1]}\t{texts[i]}\t{ciphers[i]}')




Raw key 1	Raw key 2	PlainText	CipherText
1000101110	0110101110	11010111	10111001
1000101110	0110101110	10101010	11100100
1111111111	1111111111	00000000	11101011
0000000000	0000000000	01010010	10000000
1000010110	0110101110	11111100	11100110
1011101111	0110101110	01001111	01010000
1111111111	1111111111	10101010	00000100
0000000000	0000000000	00000000	11110000


## Task2.
Implement a class called TripleSDES and complete thefollowing table:

| Raw Key  1  | Raw Key 2  | PlainText |CipherText |
| ---        |:---:       |:---:       |:---:       |
| 1000101110 | 0110101110  | 11010111 |    10111001     |
| 1000101110 | 0110101110  | 10101010 |    11100100     |
| 1111111111 | 1111111111  | 00000000 |    11101011     |
| 0000000000 | 0000000000  | 01010010 |    10000000     |
| 1000010110 | 0110101110  | 11111100 |    11100110     |
| 1011101111 | 0110101110  | 01001111 |    01010000     |
| 1111111111 | 1111111111  | 10101010 |    00000100     |
| 0000000000 | 0000000000  | 00000000 |    11110000     |

---


## Task3.
#### Cracking SDES and TripleSDES

* The message in the file cxt1.txt was encoded using SDES. Decrypt it, and find the 10-bit raw
key used for its encryption.

* The mesage in the file cxt2.txt was encoded using TripleSDES. Decrypt it, and find the two 10-
bit raw keys used for its encryption.

**Hints**: The ciphertexts are obtained by encrypting the binary string converted from clear message with
the standard ASCII code.


In [7]:
def readFile(path):
    with  open(path) as txt1:
        txt = ''.join(txt1.readlines())
        txt = txt.replace('\n','')
        txt_list = sdes.split_string(txt,8)
    return txt_list



# Le'ts try to find the suitable cipher by tyring to crack the code given the most frequent words.
eng_word_freq = ['E','T','A','O','I','N','S','R','H','D','L','U','C','M','F','Y','W','G','P','B','V','K','X','Q','J','Z']
    

Let's assume that the first 4 top bytes are somehow representing the letter 'e' or 'E', therefore we can run a brute-force algorithm to try to find the potential keys which given such letter give the expected result.

In [8]:
sdes = SDES()
txt = readFile('ctx1.txt')
chunks = sdes.get_chunk_text_frequency(txt)
print(chunks)

{'00000001': 8, '01101110': 8, '01000111': 6, '01001111': 5, '10101111': 5, '10001000': 4, '01110100': 3, '01010111': 3, '10111010': 3, '01001100': 3, '10010111': 3, '11001101': 2, '10010000': 2, '01000000': 1, '11001011': 1, '00001001': 1, '01001010': 1, '00110010': 1}


In [10]:

start_time = time.time()
e_ascii = ['e','E']
e_bin = [sdes.int2bin(ord(e),8) for e in e_ascii]
 #['00000001', '01101110', '01000111', '01001111', '10101111']
candidate_e = list(chunks.keys())[:5]
last_key = 1024 # int('1111111111',2)
decipher_text = {}
for key in range(last_key):
    key_bin = sdes.int2bin(key,10)
    for c_e in candidate_e:
        decipher_e = sdes.decrypt(c_e,key_bin)

        if decipher_e in e_bin:
            e = chr(int(decipher_e,2))
            dt = []
            is_candidate =  True
            for txt_item in txt:
                decipher = sdes.decrypt(txt_item,key_bin)
                decipher_int = int(decipher,2)
                # let's assume that encoded text consists of pure alphabetical ascii letters from 'A'(65) TO 'z' 122B
                if decipher_int > 122:
                    is_candidate = False
                decipher_chr = chr(decipher_int)
                dt.append(decipher_chr)
            if is_candidate:
                decipher_text[key_bin] = ''.join(dt).replace('\n',' ')
                print(f'Candidate key {key_bin}  => {decipher_text[key_bin]}')
    key +=1
end_time = time.time()
print(f'Elapsed time {end_time - start_time} seconds')


Candidate key 1111101010  => simplifieddesisnotsecureenoughtoprovideyousufficientsecurity
Elapsed time 0.3178083896636963


Key found for txt1 : **1111101110**  decrypted text: 

> simplifieddesisnotsecureenoughtoprovideyousufficientsecurity

#### Trying now for ctx2.txt

In [11]:
txt = readFile('ctx2.txt')
chunks = sdes.get_chunk_text_frequency(txt)
print(chunks)
start_time = time.time() 
e_ascii = ['e','E']
e_bin = [sdes.int2bin(ord(e),8) for e in e_ascii]
 #['00000001', '01101110', '01000111', '01001111', '10101111']
candidate_e = list(chunks.keys())[:5]
last_key = 1024 # int('1111111111',2)
decipher_text = {}
for k1 in range(last_key):
    k1_bin = sdes.int2bin(k1,10)
    for k2 in range(last_key):
        k2_bin = sdes.int2bin(k2,10)
        for c_e in candidate_e:
            decipher_e = sdes.decrypt3SDES(c_e,k1_bin,k2_bin)

            if decipher_e in e_bin:
                is_candidate = True
                e = chr(int(decipher_e,2))
                dt = []
                for txt_item in txt:
                    decipher = sdes.decrypt3SDES(txt_item,k1_bin,k2_bin)
                    # let's assume that encoded text consists of pure alphabetical ascii letters from 'A'(65) TO 'z' 122B
                    decipher_int = int(decipher,2)
                    if decipher_int > 122:
                        is_candidate = False
                        break
                    decipher_chr = chr(decipher_int)
                    dt.append(decipher_chr)
                if is_candidate:
                    decipher_text[(k1_bin,k2_bin)] = ''.join(dt).replace('\n',' ')
                    print(f'Candidate key [{k1_bin},{k2_bin}]  => {decipher_text[(k1_bin,k2_bin)]}')
        k2+=1
    k1 +=1
end_time = time.time()
print(f'Elapsed time {end_time - start_time} seconds')

{'10100111': 8, '10011100': 8, '00000001': 6, '10100001': 5, '01111110': 5, '11011010': 4, '11010111': 3, '01110100': 3, '10011001': 3, '11101111': 3, '00100100': 3, '11000110': 2, '01000001': 2, '00110010': 1, '01100100': 1, '10100000': 1, '10110011': 1, '00100011': 1}
Candidate key [1111101010,0101011111]  => simplifieddesisnotsecureenoughtoprovideyousufficientsecurity
Elapsed time 636.6930632591248 seconds


Key found for txt1 :  **\[ 1111101010,0101011111 \]** decrypted text:

>> simplifieddesisnotsecureenoughtoprovideyousufficientsecurity

## Task4.

Create a simple webserver which already stores two Raw keys (1000101110 - 0110101110) and can decrypt any ciphertext coming to it by that key using TripleSDES which you have already created and show the result in the browser. This is a simple end to end encryption. How strong is the security in this type of communication?

**Example:**

Browser input (bits are just for demo) :
http://localhost:5000/index.js?cipher=1011011010111011111100101111011100010111
Output : Hello

1011011010111011111100101111011100010111

0101110101010111110111001101110001010100


# To visualize the server and its functionality run:
* pip install -r example-requirements.txt
* python Server.py

### Since the keys are hidden on the server side, and if the server takes proper care of the security, the encryption algorithm might be slightly good as long as there is no clue about the algorithm used to encrypt neither the decryption algorithm. Since bot implementation (encryption and decryption) mechanisms have been done, then the strength of the algorithm will be weak since it is easy to get the keys by brute force cryptanalysis or any other techniques.
